### PR TITLE
[WIP] Feature/caliper timing

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
This needs to be spread through the other tests, but this is roughly what my changes look like. If RAJA_ENABLE_CALIPER is off, this looks an *awful* lot like nothing happened to the code. Can either do this incrementally over multiple PR's, or I can get also convert over @ajkunen 's timer stuff in Kripke (should be easier) to use this as well in this one. Goal is to have all tests outputting Caliper files when they're told to for SPOT integration

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
Mainly looking for comments from @rhornung67 and @ajkunen , on the butcheries of the RAJA/test/include timing library and Kripke Timing class respectively. If it looks good, I'll do it throughout the repo

File: cmake/SetupPackages.cmake [hunk 1]
```diff
@@ -62,6 +62,19 @@ if (RAJA_ENABLE_CUDA)
...
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DRAJA_USE_CALIPER")
+  endif()
+endif()
+   
+
 #Used for timing
 find_library(RT_LIBRARIES rt)
+if (CALIPER_FOUND)
+  list(APPEND RT_LIBRARIES ${caliper_LIB_DIR}/libcaliper.so)
+endif()
```
Comment (DavidPoliakoff) [inline cmake/SetupPackages.cmake]: @davidbeckingsale build changes okay? If Caliper is enabled, use it as a timer library

File: test/Kripke-v1.1/Kripke-v1.1-RAJA/Kripke/Timing.cpp [hunk 2]
```diff
@@ -108,6 +114,9 @@ void Timing::stop(std::string const &name){
     HPM_Stop(name.c_str());
 #endif
 
+#ifdef RAJA_USE_CALIPER
+    cali::Annotation(name.c_str()).end();
+#endif
```
Comment (DavidPoliakoff) [inline test/Kripke-v1.1/Kripke-v1.1-RAJA/Kripke/Timing.cpp]: @ajkunen this is the change to your Timing class I'm talking about

File: test/LULESH-v1.0/LULESH-v1.0_RAJA-variants/luleshRAJA-parallel-FT.cxx [hunk 3]
```diff
@@ -2756,7 +2756,7 @@ int main(int argc, char *argv[])
    RAJA::Timer timer_main;
    RAJA::Timer timer_cycle;
 
-   timer_main.start();
+   timer_main.start("timer_main");
```
Comment (DavidPoliakoff) [inline test/LULESH-v1.0/LULESH-v1.0_RAJA-variants/luleshRAJA-parallel-FT.cxx]: @rhornung67 this is the change that is pervasive, these functions taking in a const char*. If you're not using Caliper, this just calls an ordinary start/stop, otherwise it uses Caliper, which takes advantage of the name info

File: test/include/Timer.hxx [hunk 4]
```diff
@@ -93,6 +95,22 @@ public:
...
+    }
+#else
+    void start(const char* name){
+        start();
+    }
+    void stop(const char* name){
+        stop();
+    }
+#endif
+
```
Comment (DavidPoliakoff) [inline test/include/Timer.hxx]: @rhornung67 this is the change to your actual timing class

File: test/Kripke-v1.1/Kripke-v1.1-RAJA/Kripke/Timing.cpp (hunk 2 above)
Comment (ajkunen) [inline test/Kripke-v1.1/Kripke-v1.1-RAJA/Kripke/Timing.cpp]: That looks like what I would have done

**Comment by ajkunen** [comment]
I approve of the Kripke changes

File: cmake/SetupPackages.cmake (hunk 1 above)
Comment (davidbeckingsale) [inline cmake/SetupPackages.cmake]: Looks good to me. You could maybe not bother with `find_library(RT_LIBRARIES rt)` if we are using Caliper?

File: test/Kripke-v1.1/Kripke-v1.1-RAJA/Kripke/Timing.cpp (hunk 2 above)
Comment (rhornung67) [inline test/Kripke-v1.1/Kripke-v1.1-RAJA/Kripke/Timing.cpp]: Would it be better if we centralize on one timer class?  Or, maybe it won't matter since Kripke will be pulled out into a different repo.  Never  mind.

File: test/LULESH-v1.0/LULESH-v1.0_RAJA-variants/luleshRAJA-parallel-FT.cxx (hunk 3 above)
Comment (rhornung67) [inline test/LULESH-v1.0/LULESH-v1.0_RAJA-variants/luleshRAJA-parallel-FT.cxx]: I'm good with this.

File: test/include/Timer.hxx (hunk 4 above)
Comment (rhornung67) [inline test/include/Timer.hxx]: Looks good.

**Comment by rhornung67** [comment]
I'll do Adam one better. I approve the whole thing!

**Comment by ajkunen** [comment]
Oh snap!

File: test/Kripke-v1.1/Kripke-v1.1-RAJA/Kripke/Timing.cpp (hunk 2 above)
Comment (DavidPoliakoff) [inline test/Kripke-v1.1/Kripke-v1.1-RAJA/Kripke/Timing.cpp]: Yeah, this is orthogonal to the timing in RAJA itself, this is mainly meant to just work in the given applications so we can get numbers as we develop. Once RAJA has timings we should just get rid of this stuff (at least not use it for the purpose of continuous RAJA integration). More tomorrow, out for the day

**Comment by DavidPoliakoff** [comment]
If it passes CI, I see this as ready to merge. I've tested with RAJA_ENABLE_CALIPER on and off, it works either way, and all tests produce (minimally instrumented) Caliper profiles

File: cmake/SetupPackages.cmake (hunk 1 above)
Comment (DavidPoliakoff) [inline cmake/SetupPackages.cmake]: Think I'll leave it as is, want to make sure that people have the option to do both (this matters in the example of LULESH which has its own custom output for timing, somebody might want Caliper profiles AND that output)

Comment (rhornung67) [inline cmake/SetupPackages.cmake]: It's OK for now.  We'll likely change the timer class based on Tom's suggestions a few days ago.
